apps/aeutils/src/aeu_mp_trees.erl

```diff
@@ -0,0 +1,543 @@
+%%%-------------------------------------------------------------------
+%%% @copyright (C) 2017, Aeternity Anstalt
+%%% @doc Merkle Patricia Trees
+%%%
+%%% @end
+%%%-------------------------------------------------------------------
+
+-module(aeu_mp_trees).
+
+-export([ new/0
+        , new/2
+        , delete/2
+        , get/2
+        , pp/1
+        , put/3
+        , root_hash/1
+        ]).
+
+-export_type([tree/0
+             ]).
+
+-record(mpt, { hash = <<>> :: <<>> | hash() %% <<>> for the empty tree
+             , db   = dict:new() :: dict:dict()
+             }).
+
+-opaque tree() :: #mpt{}.
+
+-type tree_node() :: null() | leaf() | extension() | branch().
+
+-type null()         :: <<>>.
```

Comment:
The null() type is not used in for example tree() (mpt record). Is that on purpose? Just wondering whether the <<>> in the tree is actually also a null type. In other words, if we were to rename null() to atom null instead, would type checker complain?

Reply:
The null node is very special. It is implicitly present in all parts of the tree, it contains the null value etc. It is encoded as <<>>, so setting the type to 'null' would not work unless we used the value 'null' as well, which would not work since 'null' is not encodable in RLP.

Reply:
The tree() type is just the top level wrapper of the ADT. It is not a tree in the internal encoding of the tree. It is unwrapped/wrapped at the API functions. The internal tree has the type tree_node() which is including the null() type.